topcoffea/modules/corrections.py

```diff
@@ -285,3 +289,38 @@ def GetPUSF(nTrueInt, year, var=0):
 # test
 #val = evaluator['MuonTightSF_2016'](np.array([1.2, 0.3]),np.array([24.5, 51.3]))
 #print('val = ', val)
+
+###### Muon Rochester corrections
+################################################################
+# https://gitlab.cern.ch/akhukhun/roccor
+# https://github.com/CoffeaTeam/coffea/blob/master/coffea/lookup_tools/rochester_lookup.py
+if year=='2016': rochester_data = txt_converters.convert_rochester_file(topcoffea_path("data/MuonScale/RoccoR2016aUL.txt"), loaduncs=True)
+elif year=='2016APV': rochester_data = txt_converters.convert_rochester_file(topcoffea_path("data/MuonScale/RoccoR2016bUL.txt"), loaduncs=True)
+elif year=='2017': rochester_data = txt_converters.convert_rochester_file(topcoffea_path("data/MuonScale/RoccoR2017UL.txt"), loaduncs=True)
+elif year=='2018': rochester_data = txt_converters.convert_rochester_file(topcoffea_path("data/MuonScale/RoccoR2018UL.txt"), loaduncs=True)
+rochester = rochester_lookup.rochester_lookup(rochester_data)
+def apply_roccor(mu, is_data, var=0):
+    if not is_data:
+        hasgen = ~np.isnan(ak.fill_none(mu.matched_gen.pt, np.nan))
+        mc_rand = np.random.rand(*ak.to_numpy(ak.flatten(mu.pt)).shape)
+        mc_rand = ak.unflatten(mc_rand, ak.num(mu.pt, axis=1))
+        corrections = np.array(ak.flatten(ak.ones_like(mu.pt)))
+        errors = np.array(ak.flatten(ak.ones_like(mu.pt)))
+        
+        mc_kspread = rochester.kSpreadMC(mu.charge[hasgen],mu.pt[hasgen],mu.eta[hasgen],mu.phi[hasgen],mu.matched_gen.pt[hasgen])
+        mc_ksmear = rochester.kSmearMC(mu.charge[~hasgen],mu.pt[~hasgen],mu.eta[~hasgen],mu.phi[~hasgen],mu.nTrackerLayers[~hasgen],mc_rand[~hasgen])
+        errspread = rochester.kSpreadMCerror(mu.charge[hasgen],mu.pt[hasgen],mu.eta[hasgen],mu.phi[hasgen],mu.matched_gen.pt[hasgen])
+        errsmear = rochester.kSmearMCerror(mu.charge[~hasgen],mu.pt[~hasgen],mu.eta[~hasgen],mu.phi[~hasgen],mu.nTrackerLayers[~hasgen],mc_rand[~hasgen])
+        hasgen_flat = np.array(ak.flatten(hasgen))
+        corrections[hasgen_flat] = np.array(ak.flatten(mc_kspread))
+        corrections[~hasgen_flat] = np.array(ak.flatten(mc_ksmear))
+        errors[hasgen_flat] = np.array(ak.flatten(errspread))
+        errors[~hasgen_flat] = np.array(ak.flatten(errsmear))
+        corrections = ak.unflatten(corrections, ak.num(mu.pt, axis=1))
+        errors = ak.unflatten(errors, ak.num(mu.pt, axis=1))
+    else:
+        corrections = rochester.kScaleDT(mu.charge, mu.pt, mu.eta, mu.phi)
+        errors = rochester.kScaleDTerror(mu.charge, mu.pt, mu.eta, mu.phi)
+    if var==0: return(mu.pt * corrections) #nominal
+    elif var==1: return(mu.pt * corrections + mu.pt * errors) #up 
+    elif var==-1: return(mu.pt * corrections - mu.pt * errors) #down
```

Comment:
Should we pick a more descriptive naming scheme for the variations, say `nominal`, `up`, `down` instead of `0`, `1`, `-1`?